### DISCUSSION TITLE
contributing/development/compiling/compiling_with_dotnet

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#201] [User-contributed notes]
# contributing/development/compiling/compiling_with_dotnet

Requirements:.NET SDK 8.0+ You can use dotnet --info to check which .NET SDK versions are installed.. Enable the .NET module: By default, the .NET module is disabled when building. To enable it, ad...

https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_with_dotnet.html

**Comment by poglad-ds** [comment]
Tips:
- you actually cannot build properly without `--push-nupkgs-local` as of 4.2 and 4.3, example is not exact in that aspect
- if you tired of `./modules/mono/build_scripts/build_assemblies.py` always closing after build, just run it as `python ./modules/mono/build_scripts/build_assemblies.py` command instead. Requires python 3.6.5+

- just configure PowerShell script to run it from visual studio/rider, dont waste your own time
-> in that case, `Start-Process bin/godot.windows.editor.x86_64.mono.exe -Wait -ArgumentList '--headless --generate-mono-glue modules/mono/glue'` would be good choise as second step

**Comment by JonPetitta** [comment]
Something else not made clear is that the glue you make has to be from the exe you are going to build.  Though I think it only comes up cause I did a double build.  Was getting marshaling errors.  Anyways, the solution.  You can do a prebuild with the following scons build flag.

mono_glue=no

With that exe, you can then generate the glue, and build it with build_assemblies.py.  Then build the exe again, but leave out the above flag.  Now I finally have a mono double build of Godot that doesn't crash or throw exceptions.

**Reply by maxkoshe5** [reply]
Thank you for this comment. Was struggling with this.

**Comment by default-kramer** [comment]
When I run my exported project, I get a "unable to find the .NET assemblies directory" message. This seems to indicate a problem with the "data directory" mentioned on this page. It's not clear if this "data directory" should automatically be created by the export process, or if I need to manually copy some files from bin/GodotSharp, or what...?

**Reply by default-kramer** [reply]
Actually, maybe it is clear: "The data directory for exported projects is generated by the editor during the export." So I should see my data directory next to the .exe that is generated, but I don't. I guess I'm using the wrong export template or I built it wrong or something.

**Reply by thygrrr** [reply]
Issue is also with how the .nupkgs don't contain everything expected, but the Api lib does. This is worse in the 4.6.0-beta builds, needs to be resolved.

You can find the full files in godot/bin/GodotSharp/Api/...., and you just need to massage them into the format that .net expects to find them in in your published / outputdir.
